src/chain/chain.rs

```diff
@@ -371,9 +371,11 @@ impl Chain {
         // Check the lowest version of our info that any neighbour has and remove everything less
         // than it
         let our_oldest_ver = self.their_knowledge.values().min().map_or(0, |&v| v);
-        // TODO since the list is sorted according to the version, choose something more efficient
-        self.our_infos
-            .retain(|&(ref si, _)| *si.version() >= our_oldest_ver);
+        let oldest_ver_index = self
+            .our_infos
+            .binary_search_by_key(&our_oldest_ver, |&(ref si, _)| *si.version())
+            .unwrap_or(0);
```

Comment:
Better would be to use `unwrap_or_else(|index| index)`, to cover the cases where no entry in `our_infos` has its `version` equal to `our_oldest_ver`. Although such case might be impossible in practice, I'd still prefer bit more defensive approach.

Reply:
Isn't `unwrap_or(0)` (if in doubt, drop nothing) safer than `unwrap_or_else(|index| index)`? Maybe we should even `log_or_panic!` if `Err(i)` with `i > 0` is returned?

Reply:
Agree that it's safer to use 0 and drop nothing. I've added a `log_or_panic!` for this case.

Reply:
Update: While debugging a test failure I noticed that, in the tests, the `their_knowledge` map held by `chain` is always empty. I believe this is because `NeighbourInfo` messages are currently not handled: https://github.com/maidsafe/routing/blob/096e7c88d2136d4011ca3e5350e90eb7345b4957/src/states/elder/mod.rs#L711
Is this something that needs to be addressed? I'm happy to help with this, if so.

Reply:
its actually the `NeighbourConfirm` rpc which would be updating their_knowledge.

`NeighbourInfo` is just an empty(well with a digest to not get filtered) RoutingMessage rpc used to explicitly notify a neighbour of churn to update their neighbour_infos. Its handling(generic to any other rpc) occurs at the signed_msg layer to discover a neighbours new SectionInfo(the portion of code in your previous PR where the TODO was for the extra clone). 

As for NeighbourConfirm, while its handling code exists `handle_neighbour_confirm`, the rpc is currently not sent out when handling a new sec_info event intentionally as that would affect unack_msgs potentially not being able to get resent. With the msg_relay changes getting discussed/ongoing, once that gets into code, unacked_msgs shouldnt exist and this RPC should be ok to get enabled to be dispatched and have their_knowledge active.